### PR TITLE
Use toolchain configuration to provide an authenticated container_pull

### PR DESCRIPTION
**Pull request by smukherj1**

### CODE REVIEW

File: container/pull.bzl [hunk 1]
```diff
@@ -42,6 +42,10 @@ _container_pull_attrs = {
     "architecture": attr.string(default = "amd64"),
     "cpu_variant": attr.string(),
     "platform_features": attr.string_list(),
+    "docker_client_config": attr.string(
+        doc = "A custom directory for the docker client config.json. If left unspecified, the value of the DOCKER_CONFIG environment variable will be used. DOCKER_CONFIG is not defined, the home directory will be used.",
```
Comment (nlopezgi) [inline container/pull.bzl]: wrap line
also:
If DOCKER_CONFIG is not...

File: testing/e2e.sh [hunk 2]
```diff
@@ -423,11 +433,32 @@ EOF
...
+  cd "${ROOT}/testing/custom_toolchain_auth"
+  bazel_opts=" --override_repository=io_bazel_rules_docker=${ROOT}"
+  # Remove the old image if it exists
+  docker rmi bazel/image:image || true
+  # Push the locally built container to the private repo
+  bazel run $bazel_opts @io_bazel_rules_docker//tests/docker:push_test
+  echo "Attempting authenticated container pull and push..."
+  EXPECT_CONTAINS "$(bazel run $bazel_opts @local_pull//image)" "Loaded image"
+
+  # Run the container_pull test in the Bazel workspace that uses the default
```
Comment (nlopezgi) [inline testing/e2e.sh]: WORKSPACE

File: toolchains/docker/pull.bzl.tpl [hunk 3]
```diff
@@ -0,0 +1,12 @@
+load("@io_bazel_rules_docker//container:pull.bzl",
+    _container_pull="container_pull")
+
+# Simply call the regular container_pull with the docker client config directory
```
Comment (nlopezgi) [inline toolchains/docker/pull.bzl.tpl]: S/ Simply call the regular container_pull with ...
/ Call container_pull with ...

File: toolchains/docker/toolchain.bzl [hunk 4]
```diff
@@ -19,7 +19,7 @@ DockerToolchainInfo = provider(
     doc = "Docker toolchain rule parameters",
     fields = {
         "tool_path": "Path to the docker executable",
-        "client_config": """"A custom directory for the docker client config.json. If left unspecified, the value of the DOCKER_CONFIG environment variable will be used. DOCKER_CONFIG is not defined, the home directory will be used.""",
+        "client_config": "A custom directory for the docker client config.json. If left unspecified, the value of the DOCKER_CONFIG environment variable will be used. DOCKER_CONFIG is not defined, the home directory will be used.",
```
Comment (nlopezgi) [inline toolchains/docker/toolchain.bzl]: wrap line
If DOCKER_CONFIG is not ...

File: toolchains/docker/toolchain.bzl [hunk 5]
```diff
@@ -42,7 +42,7 @@ docker_toolchain = rule(
         ),
         "client_config": attr.string(
             default = "",
-            doc = """"A custom directory for the docker client config.json. If left unspecified, the value of the DOCKER_CONFIG environment variable will be used. DOCKER_CONFIG is not defined, the home directory will be used.""",
+            doc = "A custom directory for the docker client config.json. If left unspecified, the value of the DOCKER_CONFIG environment variable will be used. DOCKER_CONFIG is not defined, the home directory will be used.",
```
Comment (nlopezgi) [inline toolchains/docker/toolchain.bzl]: same as above

File: toolchains/docker/toolchain.bzl [hunk 6]
```diff
@@ -63,6 +63,18 @@ def _toolchain_configure_impl(repository_ctx):
         False,
     )
 
+    # Generate a custom variant authenticated version of the repository rule
+    # container_push is a custom docker client config directory was specified.
```
Comment (nlopezgi) [inline toolchains/docker/toolchain.bzl]: if a custom

File: README.md [hunk 7]
```diff
@@ -1027,6 +1027,30 @@ container_pull(
...
 
+### container_pull (Custom client configuration)
+
+If you specified a docker client directory using the "client_config" attribute
+to the docker toolchain configuration described <a href="#setup">here</a>, you
+can use a container_pull that uses the authentication credentials from the
+specified docker client directory as follows:
+
+In `WORKSPACE`:
+
```
[the remark this answers is not in the document]
Comment (smukherj1) [reply README.md]: Done

File: README.md [hunk 8]
```diff
@@ -1173,6 +1197,17 @@ use with `container_image`'s `base` attribute.
            include CPU features such as <code>["sse4", "aes"]</code>.</p>
       </td>
     </tr>
+    <tr>
+      <td><code>docker_client_config</code></td>
+      <td>
+        <p><code>string; optional</code></p>
+        <p>Specifies the directory to look for the docker client configuration. Don't use this directly.
```
[the remark this answers is not in the document]
Comment (smukherj1) [reply README.md]: Done

File: README.md [hunk 9]
```diff
@@ -1173,6 +1197,17 @@ use with `container_image`'s `base` attribute.
...
+           Specify the docker configuration directory using a custom docker toolchain configuration. Look
+           for the client_config attribute in docker_toolchain_configure <a href="#setup">here</a> for
+           details. See <a href="#container_pull-custom-client-configuration">here</a> for an example on
+           how to use container_pull after configuring the docker toolchain</p>
```
[the remark this answers is not in the document]
Comment (smukherj1) [reply README.md]: Done

File: README.md [hunk 10]
```diff
@@ -1173,6 +1197,17 @@ use with `container_image`'s `base` attribute.
            include CPU features such as <code>["sse4", "aes"]</code>.</p>
       </td>
     </tr>
+    <tr>
+      <td><code>docker_client_config</code></td>
+      <td>
+        <p><code>string; optional</code></p>
+        <p>Specifies the directory to look for the docker client configuration. Don't use this directly.
+           Specify the docker configuration directory using a custom docker toolchain configuration. Look
+           for the client_config attribute in docker_toolchain_configure <a href="#setup">here</a> for
```
[the remark this answers is not in the document]
Comment (smukherj1) [reply README.md]: Done

File: testing/e2e.sh (hunk 2 above)
Comment (smukherj1) [reply testing/e2e.sh]: done

File: toolchains/docker/pull.bzl.tpl (hunk 3 above)
Comment (smukherj1) [reply toolchains/docker/pull.bzl.tpl]: Done

File: toolchains/docker/toolchain.bzl (hunk 4 above)
Comment (smukherj1) [reply toolchains/docker/toolchain.bzl]: Done

File: toolchains/docker/toolchain.bzl (hunk 5 above)
Comment (smukherj1) [reply toolchains/docker/toolchain.bzl]: Done

File: toolchains/docker/toolchain.bzl (hunk 6 above)
Comment (smukherj1) [reply toolchains/docker/toolchain.bzl]: Done

**Pull request merged**